# Exploring Ebay Car Sales Data


In this project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle.

The aim of this project is to clean the data and analyze the included used car listings.

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

- The dataset contains 20 columns, most of which are strings.
- Some columns have null values, but none have more than ~20% null values.
- The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.


Let's convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

In [3]:
import re

def snakeize(match):
    return '_' + chr(ord(match[0])+32)

new_columns = []

for column in autos.columns:
    new_columns.append(re.sub(r'[A-Z]', snakeize, column))
    
autos.columns = new_columns

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  year_of_registration  \
0  privat    Angebot  $5,000  control          bus                  2004   
1  privat    Angebot  $8,500  control    limousine                  1997   
2  privat    Angebot  $8,990     test    limousine                  2009   
3  privat    Angebot  $4,350  control   kleinwagen                  2007   
4  privat    Angebot  $1,350     test        kombi                  2003   

     gearbox  power_p_s   model   odometer  month_of_registration fuel_type  \
0    manuell        158  andere  150,000km                      3       lpg   
1  automatik        286     7er  150,000km                      6    benzin   
2    manuell        102    golf   70,000km                      7    benzin   
3  automatik         71  fortwo   70,000km                      6    benzin   
4    manuell          0   focus  150,000km                      7    benzin   

        brand not_repaired_damage         date_created  nr_of_pictures  \
0     peugeot                nein  2016-03-26 00:00:00               0   
1         bmw                nein  2016-04-04 00:00:00               0   
2  volkswagen                nein  2016-03-26 00:00:00               0   
3       smart                nein  2016-03-12 00:00:00               0   
4        ford                nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [4]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  year_of_registration  gearbox     power_p_s  model  \
count         44905          50000.000000    47320  50000.000000  47242   
unique            8                   NaN        2           NaN    245   
top       limousine                   NaN  manuell           NaN   golf   
freq          12859                   NaN    36993           NaN   4024   
mean            NaN           2005.073280      NaN    116.355920    NaN   
std             NaN            105.712813      NaN    209.216627    NaN   
min             NaN           1000.000000      NaN      0.000000    NaN   
25%             NaN           1999.000000      NaN     70.000000    NaN   
50%             NaN           2003.000000      NaN    105.000000    NaN   
75%             NaN           2008.000000      NaN    150.000000    NaN   
max             NaN           9999.000000      NaN  17700.000000    NaN   

         odometer  month_of_registration fuel_type       brand  \
count       50000           50000.000000     45518       50000   
unique         13                    NaN         7          40   
top     150,000km                    NaN    benzin  volkswagen   
freq        32424                    NaN     30107       10687   
mean          NaN               5.723360       NaN         NaN   
std           NaN               3.711984       NaN         NaN   
min           NaN               0.000000       NaN         NaN   
25%           NaN               3.000000       NaN         NaN   
50%           NaN               6.000000       NaN         NaN   
75%           NaN               9.000000       NaN         NaN   
max           NaN              12.000000       NaN         NaN   

       not_repaired_damage         date_created  nr_of_pictures   postal_code  \
count                40171                50000         50000.0  50000.000000   
unique                   2                   76             NaN           NaN   
top                   nein  2016-04-03 00:00:00             NaN           NaN   
freq                 35232                 1946             NaN           NaN   
mean                   NaN                  NaN             0.0  50813.627300   
std                    NaN                  NaN             0.0  25779.747957   
min                    NaN                  NaN             0.0   1067.000000   
25%                    NaN                  NaN             0.0  30451.000000   
50%                    NaN                  NaN             0.0  49577.000000   
75%                    NaN                  NaN             0.0  71540.000000   
max                    NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

Seller and offer_type has all (or nearly all) the same value:

In [5]:
autos['seller'].unique()

array(['privat', 'gewerblich'], dtype=object)

In [6]:
autos[autos['seller']=='gewerblich']

date_crawled                                         name  \
7738  2016-03-15 18:06:22  Verkaufe_mehrere_Fahrzeuge_zum_Verschrotten   

          seller offer_type price   abtest vehicle_type  year_of_registration  \
7738  gewerblich    Angebot  $100  control        kombi                  2000   

      gearbox  power_p_s   model   odometer  month_of_registration fuel_type  \
7738  manuell          0  megane  150,000km                      8    benzin   

        brand not_repaired_damage         date_created  nr_of_pictures  \
7738  renault                 NaN  2016-03-15 00:00:00               0   

      postal_code            last_seen  
7738        65232  2016-04-06 17:15:37

In [7]:
autos['offer_type'].unique()

array(['Angebot', 'Gesuch'], dtype=object)

In [8]:
autos[autos['offer_type']=='Gesuch']

date_crawled                  name  seller offer_type price  \
17541  2016-04-03 15:48:33  Suche_VW_T5_Multivan  privat     Gesuch    $0   

      abtest vehicle_type  year_of_registration gearbox  power_p_s  \
17541   test          bus                  2005     NaN          0   

             model   odometer  month_of_registration fuel_type       brand  \
17541  transporter  150,000km                      0       NaN  volkswagen   

      not_repaired_damage         date_created  nr_of_pictures  postal_code  \
17541                 NaN  2016-04-03 00:00:00               0        29690   

                 last_seen  
17541  2016-04-05 15:16:06

nr_of_pictures is formed by only 0

In [9]:
autos['nr_of_pictures'].unique()

array([0])

### Clean data

So we can erased from the face of the galaxy, seller, offer_type and nr_of_pictures

In [10]:
autos = autos.drop(['seller', 'offer_type', 'nr_of_pictures'], axis = 1)
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price   abtest vehicle_type  year_of_registration    gearbox  power_p_s  \
0  $5,000  control          bus                  2004    manuell        158   
1  $8,500  control    limousine                  1997  automatik        286   
2  $8,990     test    limousine                  2009    manuell        102   
3  $4,350  control   kleinwagen                  2007  automatik         71   
4  $1,350     test        kombi                  2003    manuell          0   

    model   odometer  month_of_registration fuel_type       brand  \
0  andere  150,000km                      3       lpg     peugeot   
1     7er  150,000km                      6    benzin         bmw   
2    golf   70,000km                      7    benzin  volkswagen   
3  fortwo   70,000km                      6    benzin       smart   
4   focus  150,000km                      7    benzin        ford   

  not_repaired_damage         date_created  postal_code            last_seen  
0                nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1                nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2                nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3                nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4                nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

Uh! price and odometer are strings! What a shame!

In [11]:
re_express = re.compile('[^\d.]')
autos['price'] = autos['price'].str.replace(re_express, '').astype(float)
autos['odometer'] = autos['odometer'].str.replace(re_express, '').astype(float)
autos.rename({'price': 'price_$', 'odometer': 'odometer_km'}, axis=1, inplace=True)
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price_$   abtest vehicle_type  year_of_registration    gearbox  power_p_s  \
0   5000.0  control          bus                  2004    manuell        158   
1   8500.0  control    limousine                  1997  automatik        286   
2   8990.0     test    limousine                  2009    manuell        102   
3   4350.0  control   kleinwagen                  2007  automatik         71   
4   1350.0     test        kombi                  2003    manuell          0   

    model  odometer_km  month_of_registration fuel_type       brand  \
0  andere     150000.0                      3       lpg     peugeot   
1     7er     150000.0                      6    benzin         bmw   
2    golf      70000.0                      7    benzin  volkswagen   
3  fortwo      70000.0                      6    benzin       smart   
4   focus     150000.0                      7    benzin        ford   

  not_repaired_damage         date_created  postal_code            last_seen  
0                nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1                nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2                nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3                nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4                nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

In [12]:
autos['price_$'].sort_values(ascending=False).head(20)

39705    99999999.0
42221    27322222.0
39377    12345678.0
47598    12345678.0
27371    12345678.0
2897     11111111.0
24384    11111111.0
11137    10000000.0
47634     3890000.0
7814      1300000.0
22947     1234566.0
43049      999999.0
514        999999.0
37585      999990.0
36818      350000.0
14715      345000.0
34723      299000.0
35923      295000.0
12682      265000.0
47337      259000.0
Name: price_$, dtype: float64

oh c'mon... more then 350000 bucks for a second hand car? Are you batman? Wipe!

In [13]:
autos['price_$'].sort_values(ascending=True).value_counts()

0.0        1421
500.0       781
1500.0      734
2500.0      643
1200.0      639
           ... 
8240.0        1
44497.0       1
11270.0       1
9655.0        1
18430.0       1
Name: price_$, Length: 2357, dtype: int64

Ok let's say that all cars with a price over 350000 and under 1500 are bullshit

In [14]:
"""bullshit = autos[~autos['price_$'].between(1, 351000)].index
autos.drop(bullshit, axis=0, inplace=True)
autos.reset_index(inplace=True, drop=True)"""
autos = autos[autos['price_$'].between(1, 351000)]
autos.reset_index(inplace=True, drop=True)

autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
48560  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
48561  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
48562  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
48563  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
48564  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       price_$   abtest vehicle_type  year_of_registration    gearbox  \
0       5000.0  control          bus                  2004    manuell   
1       8500.0  control    limousine                  1997  automatik   
2       8990.0     test    limousine                  2009    manuell   
3       4350.0  control   kleinwagen                  2007  automatik   
4       1350.0     test        kombi                  2003    manuell   
...        ...      ...          ...                   ...        ...   
48560  24900.0  control    limousine                  2011  automatik   
48561   1980.0  control       cabrio                  1996    manuell   
48562  13200.0     test       cabrio                  2014  automatik   
48563  22900.0  control        kombi                  2013    manuell   
48564   1250.0  control    limousine                  1996    manuell   

       power_p_s   model  odometer_km  month_of_registration fuel_type  \
0            158  andere     150000.0                      3       lpg   
1            286     7er     150000.0                      6    benzin   
2            102    golf      70000.0                      7    benzin   
3             71  fortwo      70000.0                      6    benzin   
4              0   focus     150000.0                      7    benzin   
...          ...     ...          ...                    ...       ...   
48560        239      q5     100000.0                      1    diesel   
48561         75   astra     150000.0                      5    benzin   
48562         69     500       5000.0                     11    benzin   
48563        150      a3      40000.0                     11    diesel   
48564        101  vectra     150000.0                      1    benzin   

            brand not_repaired_damage         date_created  postal_code  \
0         peugeot                nein  2016-03-26 00:00:00        79588   
1             bmw                nein  2016-04-04 00:00:00        71034   
2      volkswagen                nein  2016-03-26 00:00:00        35394   
3           smart                nein  2016-03-12 00:00:00        33729   
4            ford                nein  2016-04-01 00:00:00        39218   
...           ...                 ...                  ...          ...   
48560        audi                nein  2016-03-27 00:00:00        82131   
48561        opel                nein  2016-03-28 00:00:00        44807   
48562        fiat                nein  2016-04-02 00:00:00        73430   
48563        audi                nein  2016-03-08 00:00:00        35683   
48564        opel                nein  2016-03-13 00:00:00        45897   

                 last_seen  
0      2016-04-06 06:45:54  
1      2016-04-06 14:45:08  
2      2016-04-06 20:15:37  
3      2016-03-15 03:16:28  
4      2016-04-01 14:38:50  
...                    ...  
48560  2016-04-01 13:47:40  
48561  2016-04-02 14:18:02  
48562  2016-04-04 11:47:27  
48563  2016-04-05 16:45:07  
48564  2016-04-06 21:18:48  

[48565 rows x 17 columns]

In [15]:
autos["price_$"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_$, dtype: float64

Given that eBay is an auction site, there could legitimately be items where the opening bid is \$1. We will keep the \$1 items, but remove anything above $350,000, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.

### Exploring the date columns

There are a number of columns with date information:

- `date_crawled`
- `registration_month`
- `registration_year`
- `ad_created`
- `last_seen`

These are a combination of dates that were crawled, and dates with meta-information from the crawler. The non-registration dates are stored as strings.

We'll explore each of these columns to learn more about the listings.

In [16]:
autos.loc[0:5, ['date_crawled','date_created','last_seen']]

date_crawled         date_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50
5  2016-03-21 13:47:45  2016-03-21 00:00:00  2016-04-06 09:45:21

To understand the date range wipe out the time part

In [17]:
autos.loc[:, 'date_crawled'] = autos.loc[:, 'date_crawled'].str[:10]
autos.loc[:, 'date_created'] = autos.loc[:, 'date_created'].str[:10]
autos.loc[:, 'last_seen'] = autos.loc[:, 'last_seen'].str[:10]
autos.loc[0:5, ['date_crawled','date_created','last_seen']]

date_crawled date_created   last_seen
0   2016-03-26   2016-03-26  2016-04-06
1   2016-04-04   2016-04-04  2016-04-06
2   2016-03-26   2016-03-26  2016-04-06
3   2016-03-12   2016-03-12  2016-03-15
4   2016-04-01   2016-04-01  2016-04-01
5   2016-03-21   2016-03-21  2016-04-06

In [18]:
autos['date_crawled'].value_counts(normalize=True, dropna=False).sort_index()*100

2016-03-05    2.532688
2016-03-06    1.404304
2016-03-07    3.601359
2016-03-08    3.329558
2016-03-09    3.308967
2016-03-10    3.218367
2016-03-11    3.257490
2016-03-12    3.691959
2016-03-13    1.566972
2016-03-14    3.654896
2016-03-15    3.428395
2016-03-16    2.960980
2016-03-17    3.162772
2016-03-18    1.291053
2016-03-19    3.477813
2016-03-20    3.788737
2016-03-21    3.737259
2016-03-22    3.298672
2016-03-23    3.222485
2016-03-24    2.934212
2016-03-25    3.160712
2016-03-26    3.220426
2016-03-27    3.109235
2016-03-28    3.486050
2016-03-29    3.409863
2016-03-30    3.368681
2016-03-31    3.183363
2016-04-01    3.368681
2016-04-02    3.547823
2016-04-03    3.860805
2016-04-04    3.648718
2016-04-05    1.309585
2016-04-06    0.317101
2016-04-07    0.140019
Name: date_crawled, dtype: float64

Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.

In [19]:
autos['last_seen'].value_counts(normalize=True, dropna=False).sort_index()*100

2016-03-05     0.107073
2016-03-06     0.432410
2016-03-07     0.539483
2016-03-08     0.741275
2016-03-09     0.959539
2016-03-10     1.066612
2016-03-11     1.237517
2016-03-12     2.378256
2016-03-13     0.889529
2016-03-14     1.260167
2016-03-15     1.587563
2016-03-16     1.645218
2016-03-17     2.808607
2016-03-18     0.735097
2016-03-19     1.583445
2016-03-20     2.065273
2016-03-21     2.063214
2016-03-22     2.137342
2016-03-23     1.853186
2016-03-24     1.976732
2016-03-25     1.921137
2016-03-26     1.680222
2016-03-27     1.564913
2016-03-28     2.085864
2016-03-29     2.234119
2016-03-30     2.477093
2016-03-31     2.378256
2016-04-01     2.279419
2016-04-02     2.491506
2016-04-03     2.520334
2016-04-04     2.448265
2016-04-05    12.476063
2016-04-06    22.180583
2016-04-07    13.194688
Name: last_seen, dtype: float64

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

In [20]:
autos['date_created'].value_counts(normalize=True, dropna=False).sort_index()*100

2015-06-11    0.002059
2015-08-10    0.002059
2015-09-09    0.002059
2015-11-10    0.002059
2015-12-05    0.002059
                ...   
2016-04-03    3.885514
2016-04-04    3.685782
2016-04-05    1.181921
2016-04-06    0.325337
2016-04-07    0.125605
Name: date_created, Length: 76, dtype: float64

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.

In [21]:
autos['year_of_registration'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: year_of_registration, dtype: float64

The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values. The minimum value is 1000, long before cars were invented and the maximum is 9999, many years into the future.

### Dealing with incorrect registration year data

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [22]:
autos = autos[autos['year_of_registration'].between(1900, 2016)]
autos.reset_index(inplace=True, drop=True)

In [23]:
autos['year_of_registration'].value_counts(normalize=True).sort_index().tail(60)*100

1957    0.004284
1958    0.008569
1959    0.012853
1960    0.049271
1961    0.012853
1962    0.008569
1963    0.017138
1964    0.025706
1965    0.036417
1966    0.047128
1967    0.055697
1968    0.055697
1969    0.040702
1970    0.081404
1971    0.055697
1972    0.070693
1973    0.049271
1974    0.051413
1975    0.038560
1976    0.044986
1977    0.047128
1978    0.094257
1979    0.072835
1980    0.182087
1981    0.059982
1982    0.087830
1983    0.109252
1984    0.109252
1985    0.203509
1986    0.154238
1987    0.154238
1988    0.289197
1989    0.372743
1990    0.743343
1991    0.726206
1992    0.792614
1993    0.910435
1994    1.347443
1995    2.628478
1996    2.941239
1997    4.179431
1998    5.062017
1999    6.205951
2000    6.760781
2001    5.646837
2002    5.325507
2003    5.781796
2004    5.790364
2005    6.289497
2006    5.719672
2007    4.877788
2008    4.744971
2009    4.466485
2010    3.403954
2011    3.476789
2012    2.806281
2013    1.720186
2014    1.420278
2015    0.8397

### Exploring price by brand

In [24]:
autos['brand'].describe()

count          46681
unique            40
top       volkswagen
freq            9862
Name: brand, dtype: object

In [25]:
autos['brand'].value_counts(normalize=True).head(20)*100

volkswagen        21.126368
bmw               11.004477
opel              10.758124
mercedes_benz      9.646323
audi               8.656627
ford               6.989996
renault            4.714980
peugeot            2.984083
fiat               2.564212
seat               1.827296
skoda              1.640925
nissan             1.527388
mazda              1.518819
smart              1.415994
citroen            1.400998
toyota             1.270324
hyundai            1.002549
sonstige_autos     0.981127
volvo              0.914719
mini               0.876159
Name: brand, dtype: float64

We're aggregating the firste 20 brands

In [27]:
brands = autos['brand'].value_counts().head(20)
brands_dict = {'brand': [], 'mean_price_$': [], 'freq': [], 'freq/price': []}
freq_list = []

for row in brands.iteritems():
    brand = row[0]
    freq = row[1]
    mean = int(autos[autos['brand'] == brand]['price_$'].sum() / freq)
    brands_dict['brand'].append(brand)
    brands_dict['mean_price_$'].append(mean)
    brands_dict['freq'].append(freq)
    brands_dict['freq/price'].append(freq/mean)    
    
brands_info_df = pd.DataFrame(brands_dict) 
brands_info_df.sort_values(by='mean_price_$', ascending=False)


brand  mean_price_$  freq  freq/price
17  sonstige_autos         12338   458    0.037121
19            mini         10613   409    0.038538
4             audi          9336  4041    0.432841
3    mercedes_benz          8628  4503    0.521905
1              bmw          8332  5137    0.616539
10           skoda          6368   766    0.120289
0       volkswagen          5402  9862    1.825620
16         hyundai          5365   468    0.087232
15          toyota          5167   593    0.114767
18           volvo          4946   427    0.086332
11          nissan          4743   713    0.150327
9             seat          4397   853    0.193996
12           mazda          4112   709    0.172422
14         citroen          3779   654    0.173062
5             ford          3749  3263    0.870365
13           smart          3580   661    0.184637
7          peugeot          3094  1393    0.450226
2             opel          2975  5022    1.688067
8             fiat          2813  1197    0.425524
6          renault          2474  2201    0.889652

In [35]:
brand_mean_mileage = []
for row in brands.iteritems():
    brand = row[0]
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage.append(int(mean_mileage))
    
mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
brands_info_df['mean_mileage_km'] = mean_mileage
brands_info_df

brand  mean_price_$  freq  freq/price  mean_mileage_km
0       volkswagen          5402  9862    1.825620           128707
1              bmw          8332  5137    0.616539           132572
2             opel          2975  5022    1.688067           129310
3    mercedes_benz          8628  4503    0.521905           130788
4             audi          9336  4041    0.432841           129157
5             ford          3749  3263    0.870365           124266
6          renault          2474  2201    0.889652           128071
7          peugeot          3094  1393    0.450226           127153
8             fiat          2813  1197    0.425524           117121
9             seat          4397   853    0.193996           121131
10           skoda          6368   766    0.120289           110848
11          nissan          4743   713    0.150327           118330
12           mazda          4112   709    0.172422           124464
13           smart          3580   661    0.184637            99326
14         citroen          3779   654    0.173062           119694
15          toyota          5167   593    0.114767           115944
16         hyundai          5365   468    0.087232           106442
17  sonstige_autos         12338   458    0.037121            89956
18           volvo          4946   427    0.086332           138067
19            mini         10613   409    0.038538            88105

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.